### ISSUE TITLE
The problem with loader.get_batch(split) when changing the image number of splits

### ISSUE THREAD

**Issue by kaelsunkiller** [ruotianluo/self-critical.pytorch#228]
I met a strange problem when I was trying to reset the number of images in train/val/test splits.
First, I changed the code in `Dataset.__init__(opt)` in `dataloader.py` to let the `split_ix` having 80k for training 40k for val and 0 for test as the original setting in mscoco. 
Then I ran the `eval.py` and got the error of being divided by 0 in function `language_eval` at line 79:
> mean_perplexity = sum([_['perplexity'] for _ in preds_filt]) / len(preds_filt)

then I found the predictions is empty caused by 
> ix1 = data['bounds']['it_max']
> ...
> for i in range(n - ix1):
       predictions.pop()

the `ix1` is 0 but it should be 40k;
and then it linked to the code in function Dataset.collate_func(batch, split):
> data['bounds'] = {'it_pos_now': it_pos_now, # the it_pos_now of the last sample
                          'it_max': len(self.split_ix[split]), 'wrapped': wrapped}

but I checked the class Dataloader and didn't find the logic conflict, the hyper `split` should be difined at initialization of the Dataloader, and I have specified the arg `split` to `val` in opts.
Then I test to get data of different `split` in loader.get_batch() and print the `data['bounds']['it_max']`. I found when set self.loaders['test'] initialized before the val split in the 'for loop', it would be normal to get the right length of the data split, but when set 'val' before 'test', the `it_max` of 'val' will be the same as 'test' (no meter it's 0 or other number, they are the same and not only at the first call of `next(self.iters[split])`).

It's a very strange problem to me. I just intented to generate the captions for 40k val split but found this. I don't know where I have ignored to change, and I found it only happens when evaluate. I checked the operations in train.py, and didn't find the difference.

I thought it may be caused by the multiprocessing function but didn't know how to fix. Hope you may have the answer, though I may not describe it very clearly. Thanks for your time anyway.

**Comment by ruotianluo** [comment]
I pushed some fix to the master, see if it helps.

**Comment by kaelsunkiller** [comment]
> I pushed some fix to the master, see if it helps.

I have found the change and fixed it, thanks a lot!
The solution is using functools.partial to freeze the arg `split` not lambda.
But I still have doubt. The problem should be the dynamic arg `split` in `collate_fn` (which should be frozen). When defined with lambda, the arg `split` becomes the same between `val` and `test` loader unconsciously, but not in `train`. I don't know why, if lambda dosen't freeze the args then why does `train` get the right arg? I guess it still relates to the multiprocessing issue?

**Comment by ruotianluo** [comment]
because it_max(which is the only place which is dependant on split in the collate_fn) is not used for training.

**Comment by kaelsunkiller** [comment]
I understand your saying, it_max is only used for eval, but I meant what I tested is excatly for evaluation program. I add some simple code as below to check the it_max in different `split` in get_batch():

>     def get_batch(self, split):
>       try:
>          data = next(self.iters[split])
>          it_max_train = next(self.iters['train'])['bounds']['it_max']
>          it_max_val = next(self.iters['val'])['bounds']['it_max']
>          it_max_test = next(self.iters['test'])['bounds']['it_max']
>          print(it_max_train, it_max_val, it_max_test)

and the result is like I discribed, only 'val' is wrong with the arg 'split'. But when I change the order in the loop of initialization:

>     for split in ['train', 'val', 'test']:
to
>     for split in ['train', 'test', 'val']:

it becomes in order. bit strange, but I think it's not about the logical problem of your codes. Anyway your work is excellent, appreciate for your contribution again.

**Comment by kaelsunkiller** [comment]
BTW, does the transformer_step.yml reproduce the results in paper [Normalized and Geometry-Aware Self-Attention Network
for Image Captioning](https://arxiv.org/pdf/2003.08897.pdf) or just follow its lr and decay strategy? I have read the comments in #196 and someone has repoduced the normalized structure in this paper, have you pushed it already? I didn't find it. Thanks.
